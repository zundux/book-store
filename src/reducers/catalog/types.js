export const FETCH_PRODUCTS_REQUEST = "catalog/FETCH_PRODUCTS_REQUEST";
export const FETCH_PRODUCTS_SUCCESS = "catalog/FETCH_PRODUCTS_SUCCESS";
export const FETCH_PRODUCTS_FAIL = "catalog/FETCH_PRODUCTS_FAIL";