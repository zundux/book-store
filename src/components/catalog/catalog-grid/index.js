import CatalogGrid from "./catalog-grid";

export default CatalogGrid;