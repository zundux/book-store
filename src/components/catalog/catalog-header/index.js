import CatalogHeader from "./catalog-header";

export default CatalogHeader;