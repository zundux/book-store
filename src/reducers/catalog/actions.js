import queryString from "query-string";
import { FETCH_PRODUCTS_FAIL, FETCH_PRODUCTS_REQUEST, FETCH_PRODUCTS_SUCCESS } from "./types";
import { getFetchParams } from "../../helpers";

export const bookResourceUrl = "https://www.googleapis.com/books/v1/volumes";
// get exact volume 'https://www.googleapis.com/books/v1/volumes/volumeId'

export const fetchingDefaultParams = {
  q: "man in black",
  filter: "paid-ebooks", // partial, full, free-ebooks, paid-ebooks, ebooks
  printType: "books", // all, books, magazines

  projection: "lite", // full, lite
  // or specify exact fields by scheme
  // firstLevelField(secondLevelField/thirdLevelField)
  // fields: 'items(volumeInfo/title)' get exact fields

  orderBy: "relevance", // relevance (default), newest

  startIndex: 0, // pagination offset,
  maxResults: 10 // 1-40 items per page
};

export const requestProducts = params => ({
  type: FETCH_PRODUCTS_REQUEST,
  params
});

export const receiveProductsSuccess = json => ({
  type: FETCH_PRODUCTS_SUCCESS,
  products: json.items,
  totalItems: json.totalItems
});

export const receiveProductsFail = error => ({
  type: FETCH_PRODUCTS_FAIL,
  errorMessage: `Could not fetch products. ${error.message}`
});

const getProductsUrl = (bookResourceUrl, requestParams) => {
  const queryParams = queryString.stringify(requestParams);
  return `${bookResourceUrl}?${queryParams}`;
};

export const fetchProducts = params => {
  return dispatch => {
    const requestParams = getFetchParams(fetchingDefaultParams, params);

    dispatch(requestProducts(requestParams));

    const fetchProductsUrl = getProductsUrl(bookResourceUrl, requestParams);

    return fetch(fetchProductsUrl)
      .then(response => response.json())
      .then(json => dispatch(receiveProductsSuccess(json)))
      .catch(error => dispatch(receiveProductsFail(error)));
  };
};
