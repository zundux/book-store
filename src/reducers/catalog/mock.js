export const fetchProductsParams = {
  params: {
    q: "man in black"
  }
};

export const fetchProductsData = {
  products: [
    { id: 1, title: "book 1" },
    { id: 2, title: "book 2" },
    { id: 3, title: "book 3" }
  ],
  totalItems: 100
};

export const fetchProductsDataPage2 = {
  products: [

    { id: 4, title: "book 4" },
    { id: 5, title: "book 5" },
    { id: 6, title: "book 6" }
  ],
  totalItems: 200
};

export const saveProductData = {
  title: "book 1", amount: 1
};

export const errorMessage = "fail request";

export default {
  fetchProductsParams,
  saveProductData,
  errorMessage
};