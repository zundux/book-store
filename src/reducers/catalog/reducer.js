import { FETCH_PRODUCTS_FAIL, FETCH_PRODUCTS_REQUEST, FETCH_PRODUCTS_SUCCESS } from "./types";
import * as R from "ramda";

export const initialStateCatalog = {
  products: [],

  params: null,
  totalItems: 0,

  isFetching: false,
  errorMessage: null
};

export default (state = initialStateCatalog, action) => {
  switch (action.type) {
    case FETCH_PRODUCTS_REQUEST:
      return {
        ...state,
        params: action.params,
        isFetching: true,
        errorMessage: null
      };

    case FETCH_PRODUCTS_SUCCESS:
      return R.mergeDeepRight(state, {
        products: action.products,
        totalItems: action.totalItems,
        isFetching: false,
        errorMessage: null
      });

    case FETCH_PRODUCTS_FAIL:
      return R.mergeDeepRight(state, {
        isFetching: false,
        errorMessage: action.errorMessage
      });

    default:
      return state;
  }
}