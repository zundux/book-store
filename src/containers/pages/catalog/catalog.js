import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import CatalogHeader from "../../../components/catalog/catalog-header";
import CatalogGrid from "../../../components/catalog/catalog-grid/catalog-grid";
import { addToCart } from "../../../reducers/cart";
import { fetchProducts } from "../../../reducers/catalog";
import ErrorLabel from "../../../components/error-label";
import LoadingLabel from "../../../components/loading-label";

export class Catalog extends React.Component {
  componentDidMount() {
    this.props.onFetchProducts();
  }

  render() {
    const {
      isFetching,
      products,
      totalItems,
      productsInCart,
      errorMessage,
      onAddToCart
    } = this.props;

    return (
      <div>
        {!isFetching && !errorMessage && products.length && (
          <div>
            <CatalogHeader productsInCart={productsInCart}/>
            {/* todo search bar */}
            <CatalogGrid
              products={products}
              onAddToCart={onAddToCart}
            />
            {totalItems && (
              <p>Found {totalItems} item{totalItems > 1 ? "s" : ""}</p>
            )}
            {/* todo pagination*/}
          </div>
        )}
        {!isFetching && errorMessage && (
          <ErrorLabel message={errorMessage}/>
        )}
        {isFetching && (
          <LoadingLabel message="Catalog loading, please wait"/>
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ catalog, cart }) => ({
  products: catalog.products,
  totalItems: catalog.totalItems,
  errorMessage: catalog.errorMessage,
  isFetching: catalog.isFetching,
  productsInCart: cart.products
});

const mapDispatchToProps = dispatch => bindActionCreators({
  onAddToCart: addToCart,
  onFetchProducts: fetchProducts
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Catalog);