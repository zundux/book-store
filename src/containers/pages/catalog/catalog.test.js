import React from "react";
import { Catalog } from "./catalog";

describe("Catalog page", () => {

  it("calls loadProducts when the site has been loaded at first time " +
    "and show loading label", () => {
    const products = [];
    const isFetchingProducts = true;
    const errorMessage = null;
    const totalItems = null;
    const productsInCart = [];
    const onAddToCart = () => {
    };
    const onFetchProducts = jest.fn();

    const component = shallow(
      <Catalog {...{
        products,
        isFetchingProducts,
        errorMessage,
        totalItems,
        productsInCart,
        onAddToCart,
        onFetchProducts
      }}/>
    );
    matchSnapshot(component);

    expect(onFetchProducts).toHaveBeenCalled();
    // todo with search params
    // expect(onFetchProduct).toHaveBeenCalledWith(params);
    //// todo should NOT have the pagination
    // todo should NOT have the total items label
  });

  it("should show product list ", () => {
    const products = [{ id: 1 }, { id: 2 }, { id: 3 }];
    const isFetchingProducts = false;
    const errorMessage = null;
    const totalItems = 101;
    const productsInCart = [];
    const onAddToCart = () => {
    };
    const onFetchProducts = () => {
    };

    const component = shallow(
      <Catalog {...{
        products,
        isFetchingProducts,
        errorMessage,
        totalItems,
        productsInCart,
        onAddToCart,
        onFetchProducts
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
    //// todo should have the pagination
    // todo should have the total items label
  });

  it("should show error message label if the products not found", () => {
    const products = [{ id: 1 }, { id: 2 }, { id: 3 }];
    const isFetchingProducts = false;
    const errorMessage = null;
    const totalItems = null;
    const productsInCart = [];
    const onAddToCart = () => {
    };
    const onFetchProducts = () => {
    };

    const component = shallow(
      <Catalog {...{
        products,
        isFetchingProducts,
        errorMessage,
        totalItems,
        productsInCart,
        onAddToCart,
        onFetchProducts
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
    //// todo should HAVE the pagination
    // todo should HAVE the total items label
  });
});