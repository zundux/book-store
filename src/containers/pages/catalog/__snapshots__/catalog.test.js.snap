// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog page calls loadProducts when the site has been loaded at first time and show loading label 1`] = `
<div>
  0
</div>
`;

exports[`Catalog page should show error message label if the products not found 1`] = `
<div>
  <div>
    <CatalogHeader
      productsInCart={Array []}
    />
    <CatalogGrid
      onAddToCart={[Function]}
      products={
        Array [
          Object {
            "id": 1,
          },
          Object {
            "id": 2,
          },
          Object {
            "id": 3,
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`Catalog page should show product list  1`] = `
<div>
  <div>
    <CatalogHeader
      productsInCart={Array []}
    />
    <CatalogGrid
      onAddToCart={[Function]}
      products={
        Array [
          Object {
            "id": 1,
          },
          Object {
            "id": 2,
          },
          Object {
            "id": 3,
          },
        ]
      }
    />
    <p>
      Found 
      101
       item
      s
    </p>
  </div>
</div>
`;
